doc/src/Developer_comm_ops.rst

```diff
@@ -0,0 +1,209 @@
+Communication operations
+------------------------
+
+This page describes various inter-processor communication operations
+encoded by LAMMPS, mostly in the core *Comm* class.  They are used by
+other classes to perform communication of different kinds.  These
+operations are useful to know about when writing new code for LAMMPS
+that needs to communicate data between processors.
+
+Owned and ghost atoms
+^^^^^^^^^^^^^^^^^^^^^
+
+As described on the :ref:`Developer_par_part` page, LAMMPS spatially
+decomposes the simulation domain, either in a *brick* or *tiled*
+manner.  Each processor (MPI task) owns atoms within its sub-domain
+and additionally stores ghost atoms within a cutoff distance of its
+sub-domain.
+
+As described on the :ref:`Developer_par_comm` page, the most common
+communication operations are first, *forward communication* which
+sends owned atom information from each processor to nearby processors
+to store with their ghost atoms.  And second, *reverse communication*
+which sends ghost atom information from each processor to the owning
+processor to accumulate (sum) the values with the corresponding owned
+atoms.
+
+The time-integration classes in LAMMPS invoke these operations each
+timestep via the *forward_comm()* and *reverse_comm()* methods in the
+*Comm* class.  
+
+Similarly, *Pair* style classes can invoke the *forward_comm_pair()*
+and *reverse_comm_pair()* methods in the *Comm* class to perform the
+same operations on data they store.  An example are many-body pair
+styles like the embedded-atom method (EAM) which compute intermediate
+values which need to be shared between owned and ghost atoms.  The
+*Comm* class methods perform the MPI communication for buffers of
+per-atom data.  They "call back" to the *Pair* class so it can *pack*
+or *unpack* the buffer with data the *Pair* class owns.  There are 4
+such methods that the *Pair* class must define, assuming it uses both
+forward and reverse communication:
+
+* pack_forward_comm()
+* unpack_forward_comm()
+* pack_reverse_comm()
+* unpack_reverse_comm()
+
+The arguments to these methods include the buffer and a list of atoms
+to pack or unpack.  The *Pair* class also sets the *comm_forward* and
+*comm_reverse* variables which store the number of values it will
+store in the communication buffers for each operation.  This means it
+can choose to store multiple per-atom values in the buffer if desired,
+and they will be communicated all together.
+
+The *Fix*, *Compute*, and *Dump* classes can invoke the same forward
+and reverse communication operations using these *Comm* class methods:
+
+* forward_comm_fix()
```

Comment:
@sjplimp Do you remember a particular reason why there is `forward_comm_pair(), `forward_comm_fix()` and so on instead of just `forward_comm()`?
With the exception of the `forward_comm_fix_variable()` those are all uniquely identifiable from the type of the first argument. So we could have them all named `forward_comm()` and the variable one `forward_comm_variable()`.

Reply:
implemented in #3148. The text here is updated to reflect that PR.